File: engine/matching_test.go
package engine

import "testing"

func TestFullFillSingleLevel(t *testing.T) {
	b := NewOrderBook()
	b.NewLimitOrder(NewOrderCmd{ClientOrderId: 10, Owner: 1, Side: Sell, Price: 100, Qty: 50}) // id 1 rests
	got := b.NewLimitOrder(NewOrderCmd{ClientOrderId: 20, Owner: 2, Side: Buy, Price: 100, Qty: 50})
	assertEvents(t, got, []Event{
		{Type: EvAccepted, OrderId: 2, ClientOrderId: 20, Owner: 2, Side: Buy, Price: 100, Qty: 50},
		{Type: EvTrade, Price: 100, Qty: 50, MakerOrderId: 1, TakerOrderId: 2, MakerOwner: 1, TakerOwner: 2},
		{Type: EvFilled, OrderId: 1, ClientOrderId: 10, Owner: 1, Side: Sell, Price: 100, Qty: 50, RemainingQty: 0},
		{Type: EvFilled, OrderId: 2, ClientOrderId: 20, Owner: 2, Side: Buy, Price: 100, Qty: 50, RemainingQty: 0},
		{Type: EvBookUpdate, Side: Sell, Price: 100, AggregateQty: 0},
	})
}

func TestPartialFillRemainderRests(t *testing.T) {
	b := NewOrderBook()
	b.NewLimitOrder(NewOrderCmd{ClientOrderId: 10, Owner: 1, Side: Sell, Price: 100, Qty: 30}) // id 1
	got := b.NewLimitOrder(NewOrderCmd{ClientOrderId: 20, Owner: 2, Side: Buy, Price: 100, Qty: 50})
	assertEvents(t, got, []Event{
		{Type: EvAccepted, OrderId: 2, ClientOrderId: 20, Owner: 2, Side: Buy, Price: 100, Qty: 50},
		{Type: EvTrade, Price: 100, Qty: 30, MakerOrderId: 1, TakerOrderId: 2, MakerOwner: 1, TakerOwner: 2},
		{Type: EvFilled, OrderId: 1, ClientOrderId: 10, Owner: 1, Side: Sell, Price: 100, Qty: 30, RemainingQty: 0},
		{Type: EvFilled, OrderId: 2, ClientOrderId: 20, Owner: 2, Side: Buy, Price: 100, Qty: 30, RemainingQty: 20},
		{Type: EvBookUpdate, Side: Sell, Price: 100, AggregateQty: 0},
		{Type: EvBookUpdate, Side: Buy, Price: 100, AggregateQty: 20},
	})
}

func TestTradesAtMakerPrice(t *testing.T) {
	b := NewOrderBook()
	b.NewLimitOrder(NewOrderCmd{ClientOrderId: 10, Owner: 1, Side: Sell, Price: 100, Qty: 10}) // id 1
	got := b.NewLimitOrder(NewOrderCmd{ClientOrderId: 20, Owner: 2, Side: Buy, Price: 105, Qty: 10})
	if got[1].Type != EvTrade || got[1].Price != 100 {
		t.Fatalf("expected trade at maker price 100, got %+v", got[1])
	}
}

func TestNonCrossingOrderRests(t *testing.T) {
	b := NewOrderBook()
	b.NewLimitOrder(NewOrderCmd{ClientOrderId: 10, Owner: 1, Side: Sell, Price: 101, Qty: 10})
	got := b.NewLimitOrder(NewOrderCmd{ClientOrderId: 20, Owner: 2, Side: Buy, Price: 100, Qty: 10})
	assertEvents(t, got, []Event{
		{Type: EvAccepted, OrderId: 2, ClientOrderId: 20, Owner: 2, Side: Buy, Price: 100, Qty: 10},
		{Type: EvBookUpdate, Side: Buy, Price: 100, AggregateQty: 10},
	})
}

func TestSweepsMultipleLevels(t *testing.T) {
	b := NewOrderBook()
	b.NewLimitOrder(NewOrderCmd{ClientOrderId: 10, Owner: 1, Side: Sell, Price: 100, Qty: 30}) // id 1
	b.NewLimitOrder(NewOrderCmd{ClientOrderId: 11, Owner: 1, Side: Sell, Price: 101, Qty: 40}) // id 2
	got := b.NewLimitOrder(NewOrderCmd{ClientOrderId: 20, Owner: 2, Side: Buy, Price: 101, Qty: 100})
	assertEvents(t, got, []Event{
		{Type: EvAccepted, OrderId: 3, ClientOrderId: 20, Owner: 2, Side: Buy, Price: 101, Qty: 100},
		{Type: EvTrade, Price: 100, Qty: 30, MakerOrderId: 1, TakerOrderId: 3, MakerOwner: 1, TakerOwner: 2},
		{Type: EvFilled, OrderId: 1, ClientOrderId: 10, Owner: 1, Side: Sell, Price: 100, Qty: 30, RemainingQty: 0},
		{Type: EvFilled, OrderId: 3, ClientOrderId: 20, Owner: 2, Side: Buy, Price: 100, Qty: 30, RemainingQty: 70},
		{Type: EvTrade, Price: 101, Qty: 40, MakerOrderId: 2, TakerOrderId: 3, MakerOwner: 1, TakerOwner: 2},
		{Type: EvFilled, OrderId: 2, ClientOrderId: 11, Owner: 1, Side: Sell, Price: 101, Qty: 40, RemainingQty: 0},
		{Type: EvFilled, OrderId: 3, ClientOrderId: 20, Owner: 2, Side: Buy, Price: 101, Qty: 40, RemainingQty: 30},
		{Type: EvBookUpdate, Side: Sell, Price: 100, AggregateQty: 0},
		{Type: EvBookUpdate, Side: Sell, Price: 101, AggregateQty: 0},
		{Type: EvBookUpdate, Side: Buy, Price: 101, AggregateQty: 30},
	})
}

func TestFifoPriorityWithinLevel(t *testing.T) {
	b := NewOrderBook()
	b.NewLimitOrder(NewOrderCmd{ClientOrderId: 10, Owner: 1, Side: Sell, Price: 100, Qty: 30}) // id 1, first
	b.NewLimitOrder(NewOrderCmd{ClientOrderId: 11, Owner: 3, Side: Sell, Price: 100, Qty: 40}) // id 2, second
	got := b.NewLimitOrder(NewOrderCmd{ClientOrderId: 20, Owner: 2, Side: Buy, Price: 100, Qty: 50})
	// id 1 must fill completely before id 2 trades at all.
	assertEvents(t, got, []Event{
		{Type: EvAccepted, OrderId: 3, ClientOrderId: 20, Owner: 2, Side: Buy, Price: 100, Qty: 50},
		{Type: EvTrade, Price: 100, Qty: 30, MakerOrderId: 1, TakerOrderId: 3, MakerOwner: 1, TakerOwner: 2},
		{Type: EvFilled, OrderId: 1, ClientOrderId: 10, Owner: 1, Side: Sell, Price: 100, Qty: 30, RemainingQty: 0},
		{Type: EvFilled, OrderId: 3, ClientOrderId: 20, Owner: 2, Side: Buy, Price: 100, Qty: 30, RemainingQty: 20},
		{Type: EvTrade, Price: 100, Qty: 20, MakerOrderId: 2, TakerOrderId: 3, MakerOwner: 3, TakerOwner: 2},
		{Type: EvFilled, OrderId: 2, ClientOrderId: 11, Owner: 3, Side: Sell, Price: 100, Qty: 20, RemainingQty: 20},
		{Type: EvFilled, OrderId: 3, ClientOrderId: 20, Owner: 2, Side: Buy, Price: 100, Qty: 20, RemainingQty: 0},
		{Type: EvBookUpdate, Side: Sell, Price: 100, AggregateQty: 20},
	})
}

func TestBestPriceOrdering(t *testing.T) {
	b := NewOrderBook()
	b.NewLimitOrder(NewOrderCmd{ClientOrderId: 1, Owner: 1, Side: Sell, Price: 102, Qty: 10}) // id 1
	b.NewLimitOrder(NewOrderCmd{ClientOrderId: 2, Owner: 1, Side: Sell, Price: 100, Qty: 10}) // id 2 - better
	got := b.NewLimitOrder(NewOrderCmd{ClientOrderId: 3, Owner: 2, Side: Buy, Price: 102, Qty: 10})
	if got[1].MakerOrderId != 2 {
		t.Fatalf("expected best ask (id 2 @100) to fill first, got %+v", got[1])
	}
}

File: java/src/test/java/javamatch/engine/MatchingTest.java
package javamatch.engine;

import org.junit.jupiter.api.Test;

import java.util.List;

import static javamatch.engine.EngineAsserts.assertEvents;
import static org.junit.jupiter.api.Assertions.assertEquals;

class MatchingTest {
    @Test
    void fullFillSingleLevel() {
        OrderBook b = new OrderBook();
        b.newLimitOrder(new NewOrderCmd(10, 1, Side.SELL, 100, 50)); // id 1 rests
        List<Event> got = b.newLimitOrder(new NewOrderCmd(20, 2, Side.BUY, 100, 50));
        assertEvents(got, List.of(
            new Event(EventType.ACCEPTED).orderId(2).clientOrderId(20).owner(2).side(Side.BUY).price(100).qty(50),
            new Event(EventType.TRADE).price(100).qty(50).makerOrderId(1).takerOrderId(2).makerOwner(1).takerOwner(2),
            new Event(EventType.FILLED).orderId(1).clientOrderId(10).owner(1).side(Side.SELL).price(100).qty(50).remainingQty(0),
            new Event(EventType.FILLED).orderId(2).clientOrderId(20).owner(2).side(Side.BUY).price(100).qty(50).remainingQty(0),
            new Event(EventType.BOOK_UPDATE).side(Side.SELL).price(100).aggregateQty(0)));
    }

    @Test
    void partialFillRemainderRests() {
        OrderBook b = new OrderBook();
        b.newLimitOrder(new NewOrderCmd(10, 1, Side.SELL, 100, 30)); // id 1
        List<Event> got = b.newLimitOrder(new NewOrderCmd(20, 2, Side.BUY, 100, 50));
        assertEvents(got, List.of(
            new Event(EventType.ACCEPTED).orderId(2).clientOrderId(20).owner(2).side(Side.BUY).price(100).qty(50),
            new Event(EventType.TRADE).price(100).qty(30).makerOrderId(1).takerOrderId(2).makerOwner(1).takerOwner(2),
            new Event(EventType.FILLED).orderId(1).clientOrderId(10).owner(1).side(Side.SELL).price(100).qty(30).remainingQty(0),
            new Event(EventType.FILLED).orderId(2).clientOrderId(20).owner(2).side(Side.BUY).price(100).qty(30).remainingQty(20),
            new Event(EventType.BOOK_UPDATE).side(Side.SELL).price(100).aggregateQty(0),
            new Event(EventType.BOOK_UPDATE).side(Side.BUY).price(100).aggregateQty(20)));
    }

    @Test
    void tradesAtMakerPrice() {
        OrderBook b = new OrderBook();
        b.newLimitOrder(new NewOrderCmd(10, 1, Side.SELL, 100, 10)); // id 1
        List<Event> got = b.newLimitOrder(new NewOrderCmd(20, 2, Side.BUY, 105, 10));
        assertEquals(EventType.TRADE, got.get(1).type);
        assertEquals(100, got.get(1).price);
    }

    @Test
    void nonCrossingOrderRests() {
        OrderBook b = new OrderBook();
        b.newLimitOrder(new NewOrderCmd(10, 1, Side.SELL, 101, 10));
        List<Event> got = b.newLimitOrder(new NewOrderCmd(20, 2, Side.BUY, 100, 10));
        assertEvents(got, List.of(
            new Event(EventType.ACCEPTED).orderId(2).clientOrderId(20).owner(2).side(Side.BUY).price(100).qty(10),
            new Event(EventType.BOOK_UPDATE).side(Side.BUY).price(100).aggregateQty(10)));
    }

    @Test
    void sweepsMultipleLevels() {
        OrderBook b = new OrderBook();
        b.newLimitOrder(new NewOrderCmd(10, 1, Side.SELL, 100, 30)); // id 1
        b.newLimitOrder(new NewOrderCmd(11, 1, Side.SELL, 101, 40)); // id 2
        List<Event> got = b.newLimitOrder(new NewOrderCmd(20, 2, Side.BUY, 101, 100));
        assertEvents(got, List.of(
            new Event(EventType.ACCEPTED).orderId(3).clientOrderId(20).owner(2).side(Side.BUY).price(101).qty(100),
            new Event(EventType.TRADE).price(100).qty(30).makerOrderId(1).takerOrderId(3).makerOwner(1).takerOwner(2),
            new Event(EventType.FILLED).orderId(1).clientOrderId(10).owner(1).side(Side.SELL).price(100).qty(30).remainingQty(0),
            new Event(EventType.FILLED).orderId(3).clientOrderId(20).owner(2).side(Side.BUY).price(100).qty(30).remainingQty(70),
            new Event(EventType.TRADE).price(101).qty(40).makerOrderId(2).takerOrderId(3).makerOwner(1).takerOwner(2),
            new Event(EventType.FILLED).orderId(2).clientOrderId(11).owner(1).side(Side.SELL).price(101).qty(40).remainingQty(0),
            new Event(EventType.FILLED).orderId(3).clientOrderId(20).owner(2).side(Side.BUY).price(101).qty(40).remainingQty(30),
            new Event(EventType.BOOK_UPDATE).side(Side.SELL).price(100).aggregateQty(0),
            new Event(EventType.BOOK_UPDATE).side(Side.SELL).price(101).aggregateQty(0),
            new Event(EventType.BOOK_UPDATE).side(Side.BUY).price(101).aggregateQty(30)));
    }

    @Test
    void fifoPriorityWithinLevel() {
        OrderBook b = new OrderBook();
        b.newLimitOrder(new NewOrderCmd(10, 1, Side.SELL, 100, 30)); // id 1, first
        b.newLimitOrder(new NewOrderCmd(11, 3, Side.SELL, 100, 40)); // id 2, second
        List<Event> got = b.newLimitOrder(new NewOrderCmd(20, 2, Side.BUY, 100, 50));
        // id 1 must fill completely before id 2 trades at all.
        assertEvents(got, List.of(
            new Event(EventType.ACCEPTED).orderId(3).clientOrderId(20).owner(2).side(Side.BUY).price(100).qty(50),
            new Event(EventType.TRADE).price(100).qty(30).makerOrderId(1).takerOrderId(3).makerOwner(1).takerOwner(2),
            new Event(EventType.FILLED).orderId(1).clientOrderId(10).owner(1).side(Side.SELL).price(100).qty(30).remainingQty(0),
            new Event(EventType.FILLED).orderId(3).clientOrderId(20).owner(2).side(Side.BUY).price(100).qty(30).remainingQty(20),
            new Event(EventType.TRADE).price(100).qty(20).makerOrderId(2).takerOrderId(3).makerOwner(3).takerOwner(2),
            new Event(EventType.FILLED).orderId(2).clientOrderId(11).owner(3).side(Side.SELL).price(100).qty(20).remainingQty(20),
            new Event(EventType.FILLED).orderId(3).clientOrderId(20).owner(2).side(Side.BUY).price(100).qty(20).remainingQty(0),
            new Event(EventType.BOOK_UPDATE).side(Side.SELL).price(100).aggregateQty(20)));
    }

    @Test
    void bestPriceOrdering() {
        OrderBook b = new OrderBook();
        b.newLimitOrder(new NewOrderCmd(1, 1, Side.SELL, 102, 10)); // id 1
        b.newLimitOrder(new NewOrderCmd(2, 1, Side.SELL, 100, 10)); // id 2 - better
        List<Event> got = b.newLimitOrder(new NewOrderCmd(3, 2, Side.BUY, 102, 10));
        assertEquals(2, got.get(1).makerOrderId, "expected best ask (id 2 @100) to fill first");
    }
}
